### ISSUE TITLE
NoSuchFileException

### ISSUE THREAD

**Issue by mahdix** [casid/jte#5]
Hi,

I'm trying to run your sample in IntelliJIdea, but get this error when I run it: `Caused by: java.nio.file.NoSuchFileException: jte/hello.jte`

I have this in my pom.xml:
```
<dependency>
            <groupId>gg.jte</groupId>
            <artifactId>jte</artifactId>
            <version>1.0.0</version>
        </dependency>
...
 <plugin>
                <groupId>gg.jte</groupId>
                <artifactId>jte-maven-plugin</artifactId>
                <version>1.0.0</version>
                <configuration>
                    <sourceDirectory>${basedir}/src/main/jte</sourceDirectory>
                    <targetDirectory>${basedir}/jte-classes</targetDirectory>
                    <contentType>Html</contentType>
                </configuration>
                <executions>
                    <execution>
                        <phase>process-classes</phase>
                        <goals>
                            <goal>precompile</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```
And in the main function:
```
CodeResolver codeResolver = new DirectoryCodeResolver(Path.of("jte")); // This is the directory where your .jte files are located.
        TemplateEngine templateEngine = TemplateEngine.create(codeResolver, ContentType.Html); // Two choices: Plain or Html
        TemplateOutput output = new StringOutput();
        templateEngine.render("hello.jte", 1, output);
        System.out.println(output);
```

How does compiler find template files? What is the path `Path.of("jte")` relative to?
Are the `.jte` files source or resources?

**Comment by casid** [comment]
Hi @mahdix!

the `Path.of("jte")` is relative to the working directory of your program. In IntelliJ you can check this when you edit your run configuration under _Working Directory_.

Where did you put your jte files? In `src/main/jte`? Then you'd need to put `new DirectoryCodeResolver(Path.of("src/main/jte"))`.

You can put the template files as resources (by using `ResourceCodeResolver`) if required, but I wouldn't recommend it, since `ResourceCodeResolver` doesn't support hot reload.

Let me know if you get it to work :-)

**Comment by mahdix** [comment]
Cool! It worked!

I had to change working directory in Run config and use "src/main/jte".

But what about production? I want to compile everything into a fat jar. In that case, what will be the correct path?

**Comment by casid** [comment]
Awesome!

I think you would do something like this for production:

```java
    private static TemplateEngine createTemplateEngine() {
        if (devSystem) {
            DirectoryCodeResolver codeResolver = new DirectoryCodeResolver(Path.of("src", "main", "jte"));
            return TemplateEngine.create(codeResolver, ContentType.Html);
        } else {
            return TemplateEngine.createPrecompiled(Path.of("jte-classes"), ContentType.Html);
        }
    }
```

You'd need to copy the jte-classes directory to production as part of your deployment.

**Comment by mahdix** [comment]
Any idea on how to automatically include compiled class files in the uber-JAR created via "mvn package" command?

**Comment by casid** [comment]
Hmm, you could instruct the jte maven plugin to compile everything to target directory. In theory this should include the files in your final build.

However, I will have to implement another `createPrecompiled` method, that looks at the existing class path, instead of a directory on the filesystem.

That'd be a pretty cool feature actually :-) Gonna try and give it a shot today.

**Comment by mahdix** [comment]
I have already tried that, but those compiled class files are not included in the jar file.

This is what I have in my pom.xml for jte plugin:
```
<configuration>
                    <sourceDirectory>${basedir}/src/main/resources/templates</sourceDirectory>
                    <targetDirectory>${basedir}/target/jte-classes</targetDirectory>
                    <contentType>Html</contentType>
                </configuration>
```
But there is not "generated" package inside "gg.jte". Looks like only standard jte classes are included int he final JAR file:
![image](https://user-images.githubusercontent.com/9294049/93012071-f144bc00-f594-11ea-9ba6-71b8ab2f3560.png)

**Comment by casid** [comment]
Hey @mahdix, the `TemplateEngine` can now be created so that the application class loader is used. This way it should be possible to bundle templates into the application JAR!

At least it worked in my test setup :-) Here's a working example:
https://github.com/casid/jte/tree/master/jte-runtime-cp-test

TemplateEngine creation is even easier that way!
`TemplateEngine.createPrecompiled(ContentType.Html);`

And in pom.xml, you'd need to change the target directory to:
`<targetDirectory>${basedir}/target/classes</targetDirectory>`

Would you mind to check it out and see if this works for you? I'd then release `1.1.0` shortly after.
 
If you want to try it you'd need to change `${jte.version}` in your pom.xml to 1.1.0-SNAPSHOT and build jte locally.

To build the jte:
- Clone this repository
- Run mvn clean install in the project's root directory

Hope it works! Also, thanks for your suggestion, I really like this new way of running in production :-)

**Comment by mahdix** [comment]
Awesom!

I did what you said above:
1. Clone this repo
2. Maven install
3. Get back to my project and change all versions to 1.1.0-SNAPSHOT, also change target path of the plugin to to `target/classes`
4. Mvn package
5. `java -jar app.jar`
6. Browse the web page at http://localhost:8080

And it shows me my template! :-)

Also inside the final JAR I have generated files:
![image](https://user-images.githubusercontent.com/9294049/93014994-83f15500-f5ad-11ea-97dc-24833392e38a.png)

On a separate matter, When I commented out my previous template content and replaced it with a simple text, I got a compiler error (This was resolved when I removed comments). Not sure if I am commenting them out correctly:

This is the template:
```
<%--@import java.util.List--%>
<%--@import io.justread.api.model.Email--%>

<%--@param List<Email> inbox--%>

<%--<table>--%>
<%--@for(var email: inbox)--%>
<%--<tr>--%>
<%--    <td>${email.getSource()}</td>--%>
<%--    <td>${email.getSubject()}</td>--%>
<%--    <td>${email.getTimestamp()}</td>--%>
<%--</tr>--%>
<%--@endfor--%>
<%--</table>--%>

this is inbox jte v6
```
The error I got when I ran `mvn package`:
```
illegal start of expression
render(jteOutput, jteHtmlInterceptor, inbox--%>);
```

Looks like either something is wrong with the way I commented out the template content or jte parser, thinks that my comments are not comments and are code.

**Comment by casid** [comment]
Oh nice, thanks for the quick update! I'm glad it worked!

Yeah it looks like this should work.. I believe the parser is confused by the commented out parameter.

Thanks for reporting. Gonna write a test & fix for this and then release `1.1.0`.

**Comment by casid** [comment]
Fixed! https://github.com/casid/jte/commit/5993bca658cef5aee93478a1d69ac7477f37e339

Gonna release `1.1.0` now :-)

**Comment by casid** [comment]
@mahdix I've updated the documentation with the new way to bundle templates:
https://github.com/casid/jte/blob/master/DOCUMENTATION.md#using-the-application-class-loader-since-110

Thanks again for the suggestion and helping me to test it 👍

**Comment by casid** [comment]
@mahdix I've just released `1.2.0`. With this version it is even easier to achieve precompiled templates with application class loader:

https://github.com/casid/jte/issues/7

Just in case you're interested :-)
